source/s3_util.c

```diff
@@ -488,6 +488,86 @@ int aws_s3_parse_content_length_response_header(
     return result;
 }
 
+int aws_s3_parse_request_range_header(
+    struct aws_http_headers *request_headers,
+    bool *out_initial_message_has_start_range,
+    bool *out_initial_message_has_end_range,
+    uint64_t *out_initial_start_range,
+    uint64_t *out_initial_end_range) {
+
+    AWS_PRECONDITION(request_headers);
+    AWS_PRECONDITION(out_initial_message_has_start_range);
+    AWS_PRECONDITION(out_initial_message_has_end_range);
+    AWS_PRECONDITION(out_initial_start_range);
+    AWS_PRECONDITION(out_initial_end_range);
+
+    bool initial_message_has_start_range = false, initial_message_has_end_range = false;
+    uint64_t initial_start_range = 0, initial_end_range = 0;
+
+    struct aws_byte_cursor range_header_value;
+
+    if (aws_http_headers_get(request_headers, g_range_header_name, &range_header_value)) {
+        return aws_raise_error(AWS_ERROR_S3_INVALID_RANGE_HEADER);
+    }
+
+    struct aws_byte_cursor range_header_start = aws_byte_cursor_from_c_str("bytes=");
+
+    int result = AWS_OP_ERR;
+    /* verify bytes= */
+    if (aws_byte_cursor_starts_with(&range_header_value, &range_header_start)) {
+        aws_byte_cursor_advance(&range_header_value, range_header_start.len);
+        struct aws_byte_cursor substr = {0};
+        /* parse start range */
+        if (!aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+        if (substr.len > 0) {
+            if (aws_byte_cursor_utf8_parse_u64(substr, &initial_start_range)) {
+                goto done;
+            }
+            initial_message_has_start_range = true;
+        }
+
+        /* parse end range */
+        if (!aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+        if (substr.len > 0) {
+            if (aws_byte_cursor_utf8_parse_u64(substr, &initial_end_range)) {
+                goto done;
+            }
+            initial_message_has_end_range = true;
+        }
+
+        /* Verify that there is nothing extra */
+        if (aws_byte_cursor_next_split(&range_header_value, '-', &substr)) {
+            goto done;
+        }
+
+        /* verify that start-range <= end-range */
+        if (initial_message_has_end_range && initial_start_range > initial_end_range) {
+            goto done;
+        }
+
+        /*verify that start-range or end-range is present */
+        if (!initial_message_has_start_range && !initial_message_has_end_range) {
+            goto done;
+        }
+
+        *out_initial_message_has_start_range = initial_message_has_start_range;
+        *out_initial_message_has_end_range = initial_message_has_end_range;
+        *out_initial_start_range = initial_start_range;
+        *out_initial_end_range = initial_end_range;
+        result = AWS_OP_SUCCESS;
+    }
+
+done:
+    if (result != AWS_OP_SUCCESS) {
+        aws_raise_error(AWS_ERROR_S3_INVALID_RANGE_HEADER);
+    }
+    return result;
```

Comment:
trivial: this function doesn't need to do any cleanup ... so not sure what benefit you're getting out of this goto label.  I would do one of these:

1) just do `return aws_raise_error(AWS_ERROR_S3_INVALID_RANGE_HEADER);` everywhere, instead of `goto done`

OR 

2) simplify by just having this be an `error` label:
```
    if (something_unexpected) {
        goto error; /* more obvious than goto done */
    }

    ...
    *out_initial_message_has_start_range = initial_message_has_start_range;
    *out_initial_message_has_end_range = initial_message_has_end_range;
    *out_initial_start_range = initial_start_range;
    *out_initial_end_range = initial_end_range;
    return AWS_OP_SUCCESS;

error:
    return aws_raise_error(AWS_ERROR_S3_INVALID_RANGE_HEADER);
}

```

Reply:
Thanks, I have updated it to first option.